subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/AbstractRecompilationSpecProvider.java

```diff
@@ -16,38 +16,140 @@
+        // If the client has not set a location for SOURCE_OUTPUT, javac outputs those files to the CLASS_OUTPUT directory, so clean that instead.
+        cleanedAnyOutput |= deleteStaleFilesIn(resourcesToDelete.get(GeneratedResource.Location.SOURCE_OUTPUT), MoreObjects.firstNonNull(spec.getCompileOptions().getAnnotationProcessorGeneratedSourcesDirectory(), spec.getDestinationDir()));
+        // In the same situation with NATIVE_HEADER_OUTPUT, javac just NPEs.  Don't bother.
+        cleanedAnyOutput |= deleteStaleFilesIn(resourcesToDelete.get(GeneratedResource.Location.NATIVE_HEADER_OUTPUT), spec.getCompileOptions().getHeaderOutputDirectory());
+
+        return cleanedAnyOutput;
+    }
+
+    private Iterable<File> narrowDownSourcesToCompile(FileTree sourceTree, PatternSet sourceToCompile) {
+        return sourceTree.matching(sourceToCompile);
+    }
+
+    private static Map<GeneratedResource.Location, PatternSet> prepareResourcePatterns(Collection<GeneratedResource> staleResources, FileOperations fileOperations) {
+        Map<GeneratedResource.Location, PatternSet> resourcesByLocation = new EnumMap<>(GeneratedResource.Location.class);
+        for (GeneratedResource.Location location : GeneratedResource.Location.values()) {
+            resourcesByLocation.put(location, fileOperations.patternSet());
+        }
+        for (GeneratedResource resource : staleResources) {
+            resourcesByLocation.get(resource.getLocation()).include(resource.getPath());
+        }
+        return resourcesByLocation;
+    }
+
+    private void processOtherChanges(CurrentCompilation current, PreviousCompilation previous, RecompilationSpec spec, SourceFileClassNameConverter sourceFileClassNameConverter) {
+        if (spec.isFullRebuildNeeded()) {
+            return;
+        }
+        SourceFileChangeProcessor sourceFileChangeProcessor = new SourceFileChangeProcessor(previous);
+        for (FileChange fileChange : sourceChanges) {
+            if (spec.isFullRebuildNeeded()) {
+                return;
+            }
+            if (fileChange.getFileType() != FileType.FILE) {
+                continue;
+            }
+
+            String relativeFilePath = fileChange.getNormalizedPath();
+            Set<String> changedClasses = sourceFileClassNameConverter.getClassNames(relativeFilePath);
+            if (changedClasses.isEmpty() && !isIncrementalOnResourceChanges(current)) {
+                spec.setFullRebuildCause(rebuildClauseForChangedNonSourceFile(fileChange));
+            }
+            sourceFileChangeProcessor.processChange(changedClasses, spec);
+        }
+    }
+
+    protected abstract boolean isIncrementalOnResourceChanges(CurrentCompilation currentCompilation);
+
+    private void prepareFilePatterns(Collection<String> staleClasses, PatternSet filesToDelete, PatternSet sourceToCompile, SourceFileClassNameConverter sourceFileClassNameConverter) {
```

Comment:
By using the sourceFileClassNameConverter here, I could reduce the include patterns to an absolute minimum, which makes this part easier to understand and slightly more efficient.